src/main/scala/firrtl/FirrtlException.scala

```diff
@@ -0,0 +1,43 @@
+// See LICENSE for license details.
+
+package firrtl
+
+import scala.util.control.NoStackTrace
+
+@deprecated("External users should use either FirrtlUserException or their own hierarchy", "1.2")
+object FIRRTLException {
+  def defaultMessage(message: String, cause: Throwable) = {
+    if (message != null) {
+      message
+    } else if (cause != null) {
+      cause.toString
+    } else {
+      null
+    }
+  }
+}
+@deprecated("External users should use either FirrtlUserException or their own hierarchy", "1.2")
+class FIRRTLException(val str: String, cause: Throwable = null)
+  extends RuntimeException(FIRRTLException.defaultMessage(str, cause), cause)
+
+/** Exception indicating user error
+  *
+  * These exceptions indicate a problem due to bad input and thus do not include a stack trace.
+  * This can be extended by custom transform writers.
+  */
+class FirrtlUserException(message: String, cause: Throwable = null)
+  extends RuntimeException(message, cause) with NoStackTrace
```

Comment:
User error is something that's common across Stages. I think it makes sense to change this to `UserException` and move it under the `firrtl.options` package.

Going one step further, what should be the behavior if a user runs a `StageMain` from the command line and gets a `UserException`?

For an `OptionsException` (which I think should probably be changed going forward to a `UserException`), this is [caught by the `StageMain`, the message is printed as a "dramatic error", and you get a `System.exit(1)`](https://github.com/freechipsproject/firrtl/blob/master/src/main/scala/firrtl/options/Stage.scala#L72). I think that's the ideal behavior for a `UserException` generated from the FIRRTL fat jar. That may necessitate more verbose exception messages that tell which transform errored as you lose that information from the exception message in this case.

And for more explanation: the reason why this happens for `StageMain` and not for `Stage` is that if you're running a `StageMain` it's clear that this is a fat jar the user called from the command line. If it's a `Stage`, then it's being run by someone else (likely a different `StageMain`) and it should be the responsibility of the caller to figure out how to print this.

Reply:
Yeah I see what you mean that `OptionsException` is a `UserException`. I'm not sure if it makes sense to have only `firrtl.options.UserException` since that feels specific to arguments while `firrtl.FirrtlUserException` seems more generic (for example, erroring on a type in the FIRRTL AST doesn't feel like its exception supertype belongs in the `options` package). That being said, perhaps `firrtl.options.UserException` should extend `firrtl.FirrtlUserException`.

Also perhaps I should renaming it from `firrtl.FirrtlUserException` to `firrtl.UserException`

Reply:
I think `firrtl.options` was just named too early... one facet of this is option handling, but it's really a library for Stages and Phases and, with a few exceptions (like `AnnotationSeq`, `DiGraph`), is independent of FIRRTL.

Basically, I think a "user exception" is something that any Stage or Phase should be able to throw. Making this `firrtl.UserException` or `firrtl.options.UserException` communicates that better.